A61-IPv4-IPv6-dualstack-backends.md

```diff
@@ -0,0 +1,890 @@
+
+## Rationale
+
+Note that we will not support all parts of "Happy Eyeballs" as described
+in [RFC-8305][RFC-8305].  For example, because our resolver API does
+not provide a way to return some addresses without others, we will not
+start trying to connect before all of the DNS queries have returned.
+
+In Java and Go, pick_first is currently implemented inside the subchannel
+rather than at the LB policy layer.  In those implementations, it
+might work to implement Happy Eyeballs inside the subchannel, which
+would avoid the need to make pick_first the universal leaf policy,
+and in Go, it would avoid the need to move the health-checking code
+out of the subchannel.  However, that approach won't work for C-core,
+and we would like to take this opportunity to move toward a more
+uniform cross-language architecture.  Also, moving pick_first up
+to the LB policy layer in Java and Go will have the nice effect of
+making their backoff work per-address instead of across all addresses,
+which is what C-core does and what the (poorly specified) [connection
+backoff spec][backoff-spec] seems to have originally envisioned.
+
+## Implementation
+
+### C-core
+
+- move client-side health checking out of subchannel so that it can be
+  controlled by pick_first (https://github.com/grpc/grpc/pull/32709)
+- assume LB policies start in CONNECTING state
+  (https://github.com/grpc/grpc/pull/33009)
+- prep for outlier detection ejecting via health watch
+  (https://github.com/grpc/grpc/pull/33340)
+- move pick_first off of the subchannel_list library that it previously
+  shared with petiole policies, and add generic health watch support
+  (https://github.com/grpc/grpc/pull/34218)
+- change petiole policies to use generic health watch, and change outlier
+  detection to eject via health state instead of raw connectivity state
+  (https://github.com/grpc/grpc/pull/34222)
+- change ring_hash to delegate to pick_first
+  (https://github.com/grpc/grpc/pull/34244)
+- add endpoint_list library for petiole policies, and use it to change
+  round_robin to delegate to pick_first
+  (https://github.com/grpc/grpc/pull/34337)
+- change WRR to delegate to pick_first
+  (https://github.com/grpc/grpc/pull/34245)
+- implement happy eyeballs in pick_first
+  (https://github.com/grpc/grpc/pull/34426 and
+  https://github.com/grpc/grpc/pull/34717)
+- implement address interleaving for happy eyeballs
+  (https://github.com/grpc/grpc/pull/34615)
+- change resolver and LB policy APIs to support multiple addresses per
+  endpoint, and update most LB policies
+  (https://github.com/grpc/grpc/pull/33567)
+- support new xDS fields (https://github.com/grpc/grpc/pull/34506)
+- change outlier detection to handle multiple addresses per endpoint
+  (https://github.com/grpc/grpc/pull/34526)
+- change stateful session affinity to handle multiple addresses per endpoint
+  (https://github.com/grpc/grpc/pull/34472)
+
+### Java
+
+- change grpclb to delegate to RR or PF
```

Comment:
In Java it already has logic to delegate to RR or PF [GrpclbLoadBalancerProvider.java](https://github.com/grpc/grpc-java/blob/master/grpclb/src/main/java/io/grpc/grpclb/GrpclbLoadBalancerProvider.java).  Is this something different?

Reply:
grpclb re-implements RR and PF. It doesn't delegate:
https://github.com/grpc/grpc-java/blob/1680ada2fd71bc5f32fdf6fbd7a17ed06d93bd6e/grpclb/src/main/java/io/grpc/grpclb/GrpclbState.java#L487
https://github.com/grpc/grpc-java/blob/1680ada2fd71bc5f32fdf6fbd7a17ed06d93bd6e/grpclb/src/main/java/io/grpc/grpclb/GrpclbState.java#L867

Reply:
Yeah, that matches what I thought we'd said.  But I think we also recently agreed that Java and Go wouldn't actually need to change grpclb, since the grpclb balancers will never send multiple addresses per endpoint, so there's no reason grpclb needs to delegate to pick_first anyway.  So I'll remove this item for both Java and Go.